### ISSUE TITLE
Data collection and training pipeline for reward model

### ISSUE THREAD

**Issue by DallasCarraher** [DallasCarraher/minecraft-companion-bot#11]
## Overview

Implement a data collection and ML training pipeline to build a learned reward model that predicts skill success rates. This model will be trained on decision tick data, validated weekly, and integrated back into the bot to improve skill selection over time.

Inspired by Tesla's approach: every vehicle (bot) collects data, which trains a central model, which is deployed back to improve future decisions.

## Goals

1. **Data Collection Infrastructure** — Log every decision tick with minimal overhead
2. **Storage & Management** — Store structured training data efficiently (JSONL, Parquet, or SQLite)
3. **Feature Engineering** — Identify what features matter for predicting skill success
4. **Training Pipeline** — Automate weekly retraining of a reward model
5. **Deployment** — Integrate trained model back into decisionLoop for skill scoring
6. **Validation** — Track improvement: does the model actually help the bot succeed more?

## What Gets Collected

Each decision tick produces one training example:

**Input (features)**:
- Context: biome, time of day, weather, inventory, nearby blocks/entities, current goal, task type
- World state: position, health, hunger, status effects
- History: recent actions, success/failure of last N skills

**Action (label)**:
- Skill name + arguments

**Outcome (target)**:
- Skill result: ok/fail, execution time, resources gained/lost
- Secondary: was this the "right" choice? (inferred from follow-up actions and goal progress)

### Phase 1: Data Collection (Week 1)

**Goals**:
- Instrument decisionLoop to log every tick
- Write to structured format (JSONL on disk, queryable)
- Zero performance overhead

**Changes**:
- Create `src/learning/dataCollector.ts`
  - Records every decision tick to JSONL
  - Batched/debounced writes to disk
- Update `src/llm/decisionLoop.ts`
  - Call `dataCollector.recordDecisionTick(context, action, result)` after each skill
- Add config options: `DATA_COLLECTION_ENABLED`, `BOT_DATA_DIR`

**Acceptance criteria**:
- Running bot generates daily JSONL files with decision ticks
- Each line is valid JSON matching schema
- No measurable performance degradation

### Phase 2: Data Preparation (Week 2)

**Goals**:
- Build utilities to load, validate, and prepare data for training
- Feature engineering: extract/normalize features from raw context
- Handle missing/corrupt data gracefully

**Changes**:
- Create `src/learning/dataPrep.ts`
  - `loadDecisionTicks(dateRange)`: read and validate JSONL
  - `extractFeatures(context)`: convert to numeric features
  - `labelOutcome(result)`: convert to success label
  - `balanceSamples(ticks)`: handle class imbalance
- Create `scripts/prepare-training-data.ts`
  - CLI: `bun run prepare-training-data --from 2026-09-11 --to 2026-09-18`
  - Outputs: `train.csv`, `test.csv`

**Acceptance criteria**:
- Can load a week of decision ticks
- Features are normalized
- Train/test split is stratified
- Data is reproducible

### Phase 3: Training Pipeline (Week 3)

**Goals**:
- Build and train a lightweight reward model
- Validate on holdout test set
- Save model artifacts

**Recommended approach**: Small classifier (sklearn RandomForest or XGBoost)
- Input: context features
- Output: P(skill succeeds in this context)
- Pros: Fast, interpretable, runs on-device
- Cons: Requires feature engineering

**Changes**:
- Create `src/learning/train.ts` or `scripts/train-model.ts`
  - Loads data, trains, validates, saves model
  - Logs: accuracy, F1, confusion matrix, per-skill performance
- Create `scripts/evaluate-model.ts`
  - Compare model performance on test data

**Acceptance criteria**:
- Model trains in < 5 minutes on a week of data
- Test accuracy > 70%
- Per-skill metrics show which skills are easy/hard to predict
- Model serializes and loads correctly

### Phase 4: Model Integration (Week 4)

**Goals**:
- Load trained model in decisionLoop
- Score skills before presenting to LLM
- Rank/filter by predicted success rate

**Changes**:
- Create `src/learning/rewardModel.ts`
  - `loadModel(path)`: deserialize model
  - `scoreSkill(context, skillName, args)`: predict P(success)
- Update `src/llm/skillFilter.ts`
  - Filter out low-confidence skills (< threshold)
  - Optionally re-rank by score or include score in description
- Update `src/llm/contextBuilder.ts`
  - Include "learnedSkillScores" in DecisionContext
- Update `index.ts`
  - Load model on startup

**Acceptance criteria**:
- Model loads in < 1s at startup
- Scoring each skill takes < 10ms
- LLM sees predicted success rates in context
- Skill filtering uses model scores

### Phase 5: Validation & Monitoring (Ongoing)

**Goals**:
- Measure if the model actually improves bot performance
- Track metrics: goal success rate, average task completion time, skill failure rates
- Automate weekly retraining

**Changes**:
- Create `src/learning/metrics.ts`
  - Track: tasks completed, goals reached, average time per goal, skill success rate
- Create `scripts/weekly-retrain.ts`
  - Triggered manually or via cron
  - Collect data, prepare, train, evaluate, save if improved
- Optional: Create monitoring dashboard
  - JSON metrics file or HTML report showing improvement over time

**Acceptance criteria**:
- Metrics tracked weekly
- Retraining is automated (single command)
- Can compare two models' performance
- Logs show improvement (or regressions)

## Data Schema

```typescript
interface DecisionTickData {
  timestamp: string; // ISO 8601
  tickId: string;
  
  context: {
    biome: string;
    timeOfDay: number; // 0-24
    position: { x: number; y: number; z: number };
    health: number;
    inventory: { name: string; count: number }[];
    nearbyBlocks: { name: string; distance: number }[];
    nearbyEntities: { type: string; isHostile: boolean; distance: number }[];
    goal: string | null;
    taskType: string | null;
    previousSkills: { name: string; ok: boolean }[];
  };
  
  action: {
    skillName: string;
    args: Record<string, unknown>;
  };
  
  outcome: {
    ok: boolean;
    message: string;
    executionTimeMs: number;
    data?: Record<string, unknown>;
  };
  
  goalProgress?: 'on_track' | 'blocked' | 'completed' | 'failed';
}
```

## Storage Strategy

- **Format**: JSONL (one JSON per line, one file per day)
  - Pros: Human-readable, easy to stream
  - Cons: Requires parsing for analysis
- **Location**: `./.bot-data/decision-ticks-YYYY-MM-DD.jsonl`
- **Retention**: Keep 90 days; archive older data
- **Size estimate**: ~500 bytes per tick × 86,400 ticks/day = ~40 MB/day; ~1.2 GB/month

## ML Approach Recommendation

Start with **sklearn + feature engineering** (RandomForest or XGBoost):
- ✅ Fast to implement, no deep learning infrastructure
- ✅ Runs on bot machine (no cloud needed)
- ✅ Interpretable (can see which features matter)
- ✅ Proves the value before investing in complex models

If that works well, upgrade to fine-tuning or neural nets later.

## Dependency on Other Work

- **Skill Robustness** (parallel): Will provide richer outcome data (partial success, execution time)
- **Skill Composition** (parallel): Will help understand which sequences work
- **Progressive Filtering** (parallel): Will use trained model to filter skills before LLM call

## Out of Scope

- Real-time model updates (retrain only weekly)
- A/B testing different models
- Distributed/cloud training (start single-machine)
- Fine-tuning LLMs (focus on reward model first)
- Handling concept drift (revisit if it becomes an issue)

## References

- Architecture: docs/architecture.md
- Decision loop: src/llm/decisionLoop.ts
- Context builder: src/llm/contextBuilder.ts
- Memory store: src/memory/store.ts
